pandas/core/reshape/merge.py

```diff
@@ -787,7 +787,25 @@ def _maybe_add_join_keys(self, result, left_indexer, right_indexer):
             take_left, take_right = None, None
 
             if name in result:
-
+                array_like = is_array_like(rname) or is_array_like(lname)
```

Comment:
what are you trying to do here?

Reply:
@jreback

If you mean only line 790: If an array is given as join key, then this should be handled like a column. So we have to set the values into the result.

Generally this block should handle the following:

If we join an index with an column this part sets the index values as the column values in the result DataFrame. To avoid this I check if we should run in there. 

We should only run in there if both DataFrames are joined via the same columns, if a array was given as join condition. 

Actually I realised that the condition with the index is crap. That would keep the initial problem if the index is accidentally named as the join column. This will break a few more tests unfortunately.